File: backend/access_profiles.py
"""Least-privilege API profiles for narrowly-scoped staff accounts."""
from __future__ import annotations

from re import fullmatch

GROUND_TILE_QUOTATIONS_FOLLOWUPS = "ground_tile_quotations_followups"
GROUND_PAYMENTS_DISPATCHES = "ground_payments_dispatches"
SANITARY_QUOTATIONS_FOLLOWUPS = "sanitary_quotations_followups"
SANITARY_PURCHASES = "sanitary_purchases"

PROFILE_MODULES = {
    GROUND_TILE_QUOTATIONS_FOLLOWUPS: {"tiles", "followups"},
    GROUND_PAYMENTS_DISPATCHES: {"payments", "orders"},
    SANITARY_QUOTATIONS_FOLLOWUPS: {"quotations", "followups"},
    SANITARY_PURCHASES: {"purchases"},
}

# Account/session functions and the single assigned-floor selector. No
# business records are exposed by these endpoints.
_COMMON = {
    ("GET", "/api/auth/me"), ("POST", "/api/auth/change-password"),
    ("POST", "/api/auth/logout"), ("GET", "/api/auth/sessions"),
    ("POST", "/api/auth/sessions/logout-all"),
    ("GET", "/api/settings/floor-access"), ("GET", "/api/settings/floors"),
    ("GET", "/api/settings/permission-matrix"), ("GET", "/api/roles"),
}


def _catalog_read(method: str, path: str) -> bool:
    return method == "GET" and (
        path.startswith("/api/products") or path.startswith("/api/catalog/")
        or path in {"/api/brands", "/api/categories"}
        or fullmatch(r"/api/families/[^/]+", path) is not None
        or fullmatch(r"/api/products/[^/]+/media", path) is not None
    )


def _customer_for_quotation(method: str, path: str) -> bool:
    """Lookup/create/update a quote's customer, never use CRM-only actions."""
    if path == "/api/customers":
        return method in {"GET", "POST"}
    return bool(fullmatch(r"/api/customers/[^/]+", path) and method in {"GET", "PATCH"})


def _dispatch_surface(method: str, path: str) -> bool:
    """The complete tile-orders workspace, and no other business area.

    The ground-floor dispatch worker has full capability for tile orders.  This
    route family includes the supporting brand, inventory, history, chalan and
    release/movement calls required by that single workspace.
    """
    return path.startswith("/api/tile-orders/")


def profile_allows_request(profile: str | None, method: str, path: str) -> bool:
    """Fail-closed API allow-list. No profile preserves existing staff access."""
    if not profile:
        return True
    method = method.upper()
    if (method, path) in _COMMON or (method == "DELETE" and fullmatch(r"/api/auth/sessions/[^/]+", path)):
        return True
    if profile in {GROUND_TILE_QUOTATIONS_FOLLOWUPS, SANITARY_QUOTATIONS_FOLLOWUPS}:
        return (path.startswith("/api/quotations") or path.startswith("/api/followups")
                or _customer_for_quotation(method, path) or _catalog_read(method, path)
                or (method in {"GET", "POST"} and path == "/api/referrers")
                or (method == "POST" and path == "/api/products/custom")
                or (method == "POST" and path == "/api/downloads/token"))
    if profile == GROUND_PAYMENTS_DISPATCHES:
        return (path.startswith("/api/payments") or _dispatch_surface(method, path)
                or (method == "POST" and path == "/api/downloads/token"))
    if profile == SANITARY_PURCHASES:
        return (path.startswith("/api/purchases") or path.startswith("/api/purchase-orders")
                or path.startswith("/api/suppliers")
                or (method == "GET" and path.startswith("/api/activity/purchase/"))
                or (method == "GET" and path == "/api/customers")
                or (method == "POST" and path == "/api/downloads/token"))
    return False

File: frontend/src/access-profiles.ts
export type AccessProfile =
  | "ground_tile_quotations_followups"
  | "ground_payments_dispatches"
  | "sanitary_quotations_followups"
  | "sanitary_purchases";

export const PROFILE_PROVISIONING: Record<AccessProfile, {
  label: string;
  description: string;
  floorId: string;
  minimumRole: "sales" | "accounts" | "purchase";
}> = {
  ground_tile_quotations_followups: {
    label: "Ground Floor · Tile Quotations & Follow-ups",
    description: "Tiles, quotations, follow-ups, and read-only customer delivery status.",
    floorId: "ground-floor",
    minimumRole: "sales",
  },
  ground_payments_dispatches: {
    label: "Ground Floor · Payments & Dispatches",
    description: "Ground Floor payments and tile dispatch workflow.",
    floorId: "ground-floor",
    minimumRole: "accounts",
  },
  sanitary_quotations_followups: {
    label: "Sanitary Bathroom · Quotations & Follow-ups",
    description: "Sanitary quotations and customer follow-ups.",
    floorId: "first-floor",
    minimumRole: "sales",
  },
  sanitary_purchases: {
    label: "Sanitary Bathroom · Purchases",
    description: "Sanitary purchase orders, suppliers, and purchase tracking.",
    floorId: "first-floor",
    minimumRole: "purchase",
  },
};

export const PROFILE_MODULES: Record<AccessProfile, Set<string>> = {
  ground_tile_quotations_followups: new Set(["tiles", "followups"]),
  ground_payments_dispatches: new Set(["payments", "orders"]),
  sanitary_quotations_followups: new Set(["quotations", "followups"]),
  sanitary_purchases: new Set(["purchases"]),
};

export function staffLandingPath(profile?: AccessProfile | null): string {
  switch (profile) {
    case "ground_tile_quotations_followups": return "/(admin)/tiles";
    case "ground_payments_dispatches": return "/(admin)/payments";
    case "sanitary_quotations_followups": return "/(admin)/quotations";
    case "sanitary_purchases": return "/(admin)/purchases";
    default: return "/(admin)/dashboard";
  }
}
